src/Workspaces/CoreTestUtilities/Remote/TestSerializerService.cs

```diff
@@ -64,9 +69,47 @@ public override MetadataReference ReadMetadataReferenceFrom(ObjectReader reader,
             }
         }
 
+        public override void WriteAnalyzerReferenceTo(AnalyzerReference reference, ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            if (reference is TestGeneratorReference generatorReference)
+            {
+                // It's a test reference, we'll just store it in a map and then just write out our GUID
+                _sharedTestGeneratorReferences.TryAdd(generatorReference.Guid, generatorReference);
+                writer.WriteGuid(generatorReference.Guid);
+            }
+            else
+            {
+                writer.WriteGuid(Guid.Empty);
+                base.WriteAnalyzerReferenceTo(reference, writer, cancellationToken);
+            }
+        }
+
+        public override AnalyzerReference ReadAnalyzerReferenceFrom(ObjectReader reader, CancellationToken cancellationToken)
+        {
+            var testGeneratorReferenceGuid = reader.ReadGuid();
+
+            if (testGeneratorReferenceGuid != Guid.Empty)
+            {
+                Contract.ThrowIfFalse(_sharedTestGeneratorReferences.TryGetValue(testGeneratorReferenceGuid, out var generatorReference));
+                return generatorReference;
+            }
+            else
+            {
+                return base.ReadAnalyzerReferenceFrom(reader, cancellationToken);
+            }
+        }
+
         [ExportWorkspaceServiceFactory(typeof(ISerializerService), layer: ServiceLayer.Test), Shared, PartNotDiscoverable]
+        [Export(typeof(Factory))]
         internal new sealed class Factory : IWorkspaceServiceFactory
         {
+            /// <summary>
+            /// In unit tests that are testing OOP, we want to be able to share test generator references directly
+            /// from one side to another. We'll create multiple instances of the TestSerializerService -- one for each
+            /// workspace and in different MEF compositions, but we'll share the generator references across them.
+            /// </summary>
+            public ConcurrentDictionary<Guid, TestGeneratorReference> SharedTestGeneratorReferences { get; set; } = new();
```

Comment:
is this allowed to be called multiple times?  Wondering if it would be better to turn this into a set method that throws if it's called more than once.

And - should this throw if it has not been set at least once?

Reply:
Switched to a property that only allows a single set, and only if nobody had already read it.